### PR TITLE
Timespan division and multiplication.

### PR DESCRIPTION
**Pull request by JonHanna**
https://github.com/dotnet/corefx/issues/7175

### CODE REVIEW

**Comment by JonHanna** [comment]
Tests at dotnet/corefx#16475

**Comment by JonHanna** [comment]
Test CentOS7.1 x64 Debug Build and Test

File: src/mscorlib/src/System/TimeSpan.cs [hunk 1]
```diff
@@ -405,6 +405,42 @@ public String ToString(String format, IFormatProvider formatProvider)
...
+        {
+            double ticks = timeSpan.Ticks / divisor;
+            if (ticks > long.MaxValue || ticks < long.MinValue || double.IsNaN(ticks))
+            {
+                // Assuming exceptions are the rarer case, and since both IsNaN(divisor) and 0/0 can
+                // result in IsNaN(ticks), defer determining which exception until this branch has
+                // been taken.
+                throw double.IsNaN(divisor)
+                    ? new ArgumentException(Environment.GetResourceString("Arg_CannotBeNaN"), nameof(divisor))
+                    : (Exception)new OverflowException(Environment.GetResourceString("Overflow_TimeSpanTooLong"));
```
Comment (danmosemsft) [inline src/mscorlib/src/System/TimeSpan.cs]: Why is this cast to Exception needed?

**Comment by danmosemsft** [comment]
@tarekgh can you review also?

File: src/mscorlib/src/System/TimeSpan.cs (hunk 1 above)
Comment (joelweiss) [inline src/mscorlib/src/System/TimeSpan.cs]: > [Either the type of first_expression and second_expression must be the same, or an implicit conversion must exist from one type to the other.
> ](https://msdn.microsoft.com/en-us/library/ty67wk28.aspx)

Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: Yep, though it's one of those casts that don't have any effect on the compiled code, but are needed to make the expression valid.

File: src/mscorlib/src/System/TimeSpan.cs [hunk 2]
```diff
@@ -405,6 +405,42 @@ public String ToString(String format, IFormatProvider formatProvider)
...
         }
 
+        public static TimeSpan operator *(TimeSpan timeSpan, double factor)
+        {
+            if (double.IsNaN(factor))
+            {
+                throw new ArgumentException(Environment.GetResourceString("Arg_CannotBeNaN"), nameof(factor));
+            }
+
+            double ticks = timeSpan.Ticks * factor;
```
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: >timeSpan.Ticks * factor; [](start = 27, length = 24)

should we put this inside checked() to throw for the overflow too?

File: src/mscorlib/src/System/TimeSpan.cs [hunk 3]
```diff
@@ -405,6 +405,42 @@ public String ToString(String format, IFormatProvider formatProvider)
...
+            if (ticks > long.MaxValue | ticks < long.MinValue)
+            {
+                throw new OverflowException(Environment.GetResourceString("Overflow_TimeSpanTooLong"));
+            }
+
+            return FromTicks((long)ticks);
```
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: >(long)ticks [](start = 29, length = 11)

should we care about rounding?

File: src/mscorlib/src/System/TimeSpan.cs [hunk 4]
```diff
@@ -405,6 +405,42 @@ public String ToString(String format, IFormatProvider formatProvider)
...
+
+        public static TimeSpan operator /(TimeSpan timeSpan, double divisor)
+        {
+            double ticks = timeSpan.Ticks / divisor;
+            if (ticks > long.MaxValue || ticks < long.MinValue || double.IsNaN(ticks))
+            {
+                // Assuming exceptions are the rarer case, and since both IsNaN(divisor) and 0/0 can
+                // result in IsNaN(ticks), defer determining which exception until this branch has
+                // been taken.
+                throw double.IsNaN(divisor)
```
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: >double.IsNaN(divisor) [](start = 22, length = 21)

it is better to do this check first in the beginning of the method and then make the code cleaner. I understand you want to optimize but I think one check here is not really a big deal.

File: src/mscorlib/src/System/TimeSpan.cs [hunk 5]
```diff
@@ -405,6 +405,42 @@ public String ToString(String format, IFormatProvider formatProvider)
...
+            }
+
+            return FromTicks((long)ticks);
+        }
+
+        public static TimeSpan operator *(double factor, TimeSpan timeSpan) => timeSpan * factor;
+
+        public static TimeSpan operator /(TimeSpan timeSpan, double divisor)
+        {
+            double ticks = timeSpan.Ticks / divisor;
```
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: >timeSpan.Ticks / divisor; [](start = 27, length = 25)

should we put this inside checked()? for better determination of overflow/underflow?

File: src/mscorlib/src/System/TimeSpan.cs [hunk 6]
```diff
@@ -405,6 +405,42 @@ public String ToString(String format, IFormatProvider formatProvider)
...
+                // been taken.
+                throw double.IsNaN(divisor)
+                    ? new ArgumentException(Environment.GetResourceString("Arg_CannotBeNaN"), nameof(divisor))
+                    : (Exception)new OverflowException(Environment.GetResourceString("Overflow_TimeSpanTooLong"));
+            }
+
+            return FromTicks((long)ticks);
+        }
+
+        public static double operator /(TimeSpan t1, TimeSpan t2) => t1.Ticks / (double)t2.Ticks;
```
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: >TimeSpan t2 [](start = 53, length = 11)

do we need to allow dividing by TimeSpan(0)? ticks is really long and theoretically we are dividing by long 0 and not double 0

File: src/mscorlib/src/System/TimeSpan.cs (hunk 2 above)
Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: No effect on double arithmetic.

File: src/mscorlib/src/System/TimeSpan.cs (hunk 3 above)
Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: I think it's probably best to be consistent with `FromHours()` etc. and they do a direct cast from double to long after explicitly checking for overflow too.

File: src/mscorlib/src/System/TimeSpan.cs (hunk 4 above)
Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: Sure. Will do.

File: src/mscorlib/src/System/TimeSpan.cs (hunk 6 above)
Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: Practically we're dividing longs as a matter of representation, but theoretically we're dividing a span of time by another. I think it's reasonable to be able to ask the question "how many zero-second periods in an hour?"

File: src/mscorlib/src/System/TimeSpan.cs (hunk 3 above)
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: Actually FromHours rounds to the milliseconds,

Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: Right you are. That seems probably too coarse for this, in that `FromTicks(2) * 2.0` should probably work as one would expect.

File: src/mscorlib/src/System/TimeSpan.cs (hunk 5 above)
Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: Ignored by the compiler on double arithmetic.

File: src/mscorlib/src/System/TimeSpan.cs (hunk 2 above)
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: thanks for reminding me about it :-)

File: src/mscorlib/src/System/TimeSpan.cs (hunk 6 above)
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: I am fine with this interpenetration. could you please put some comments around the added APIs describing the design and the issues discussed here?

File: src/mscorlib/src/System/TimeSpan.cs (hunk 3 above)
Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: but what about FromTicks(2) * 2.9?

---
In reply to: [103531161](https://github.com/dotnet/coreclr/pull/9804#discussion_r103531161) [](ancestors = 103531161)

Comment (JonHanna) [inline src/mscorlib/src/System/TimeSpan.cs]: I think, on balance, that rounding to the nearest tick is probably the best approach, so I'll update to do that. It's only going to matter when half a tick makes a difference, but it's probably what would have the least surprises for users.

Comment (tarekgh) [inline src/mscorlib/src/System/TimeSpan.cs]: thanks make sense. thank you.

**Comment by tarekgh** [comment]
LGTM. thanks @JonHanna for getting this ready. are you willing to enable exposing these APIs in the corefx repo?

**Comment by stephentoub** [comment]
@tarekgh, he already created the PR: https://github.com/dotnet/corefx/pull/16475

**Comment by tarekgh** [comment]
thanks @stephentoub

**Comment by JonHanna** [comment]
Yes, it needs some fixup for uapaot but I should have that ready before the changes here are visible to corefx.

**Comment by tarekgh** [comment]
The UAP-aot changes should go to corert repo and not here.

**Pull request merged**